synapse/storage/databases/main/events.py

```diff
@@ -1773,31 +1729,53 @@ def _update_current_state_txn(
-                        # `event_stream_ordering` here is because the column has a `NON
-                        # NULL` constraint and we need *some* answer. And if the row
-                        # already exists, it already has the correct value and it's
-                        # better to just rely on
-                        # `_update_sliding_sync_tables_with_new_persisted_events_txn()`
-                        # to do the right thing (same for `bump_stamp`).
-                        "event_stream_ordering": sliding_sync_table_changes.joined_room_best_effort_most_recent_stream_ordering,
-                        # If we're trying to fully-insert a row, we need to provide a
-                        # value for `bump_stamp` if it exists for the room.
-                        "bump_stamp": sliding_sync_table_changes.joined_room_bump_stamp_to_fully_insert,
-                    },
+                args: List[Any] = [
+                    room_id,
+                    room_id,
+                    sliding_sync_table_changes.joined_room_bump_stamp_to_fully_insert,
+                ]
+                args.extend(iter(sliding_sync_updates_values))
+
+                # We use a sub-query for `stream_ordering` because it's unreliable to
+                # pre-calculate from `events_and_contexts` at the time when
+                # `_calculate_sliding_sync_table_changes()` is ran. We could be working
+                # with events that were previously persisted as an `outlier` with one
+                # `stream_ordering` but are now being persisted again and de-outliered
+                # and assigned a different `stream_ordering`. Since we call
+                # `_calculate_sliding_sync_table_changes()` before
+                # `_update_outliers_txn()` which fixes this discrepancy (always use the
+                # `stream_ordering` from the first time it was persisted), we're working
+                # with an unreliable `stream_ordering` value that will possibly be
+                # unused and not make it into the `events` table.
+                #
+                # We don't update `event_stream_ordering` `ON CONFLICT` because it's
+                # simpler and we can just rely on
+                # `_update_sliding_sync_tables_with_new_persisted_events_txn()` to do
+                # the right thing (same for `bump_stamp`). The only reason we're
+                # inserting `event_stream_ordering` here is because the column has a
+                # `NON NULL` constraint and we need some answer.
+                txn.execute(
+                    f"""
+                    INSERT INTO sliding_sync_joined_rooms
+                        (room_id, event_stream_ordering, bump_stamp, {", ".join(sliding_sync_updates_keys)})
+                    VALUES (
+                        ?,
+                        (SELECT stream_ordering FROM events WHERE room_id = ? ORDER BY stream_ordering DESC LIMIT 1),
```

Comment:
This could be simplified to remove `ORDER BY stream_ordering DESC`. This just gives a more correct answer  but we will end up with the correct answer anyway when `_update_sliding_sync_tables_with_new_persisted_events_txn()` runs right after this.

```suggestion
                        (SELECT stream_ordering FROM events WHERE room_id = ? LIMIT 1),
```

Up to you